azkaban-web-server/src/main/java/azkaban/scheduler/QuartzHacker.java

```diff
@@ -0,0 +1,141 @@
+package azkaban.scheduler;
+
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.Optional;
+import org.quartz.Scheduler;
+import org.quartz.core.QuartzSchedulerResources;
+import org.quartz.impl.jdbcjobstore.DriverDelegate;
+import org.quartz.impl.jdbcjobstore.NoSuchDelegateException;
+import org.quartz.impl.jdbcjobstore.StdJDBCDelegate;
+import org.quartz.spi.ClassLoadHelper;
+import org.quartz.spi.JobStore;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/***
+ * This class exists solely to fix https://github.com/azkaban/azkaban/issues/2461
+ *
+ * It is a necessary evil to help a minority of users affected by a bug related
+ * to serialization and database corruption.
+ *
+ * It handles unconfigured cases where the bug is encountered and allows for automated
+ * remediation measures when QuartzScheduler::enableSerializationHack is triggered
+ * by appropriate exceptions, identified by QuartzScheduler::isSerializationBug.
+ */
+public class QuartzHacker {
+
+  private static final Logger logger = LoggerFactory.getLogger(QuartzHacker.class);
+
+  static Scheduler fix(final Scheduler scheduler) {
+    org.quartz.core.QuartzScheduler quartzScheduler = getPrivateField(scheduler, "sched",
+        org.quartz.core.QuartzScheduler.class);
+    if (quartzScheduler == null) {
+      logger.info("No sched field in scheduler");
+      return null;
+    }
+    QuartzSchedulerResources resources = getPrivateField(quartzScheduler, "resources",
+        QuartzSchedulerResources.class);
+    if (resources == null) {
+      logger.info("No resources field in quartzScheduler");
+      return null;
+    }
+    JobStore jobStore = resources.getJobStore();
+    DriverDelegate delegate = getPrivateField(jobStore, "delegate", DriverDelegate.class);
+    if (delegate == null) {
+      logger.info("No resources delegate in jobStore");
+      return null;
+    }
+    if (!(delegate instanceof StdJDBCDelegate)) {
+      logger.info(String
+          .format("Class of delegate is not StdJDBCDelegate (%s)", delegate.getClass().getName()));
+      return null; // Do not re-wrap!
+    }
+    if (delegate instanceof CustomQuartzDeserializer) {
+      logger.info("Class of delegate is already CustomQuartzDeserializer");
+      return null; // Do not re-wrap!
+    }
+    StdJDBCDelegate inner = (StdJDBCDelegate) delegate;
+    CustomQuartzDeserializer value = new CustomQuartzDeserializer(inner);
```

Comment:
The additional overrides are there to make sure that ONLY the relevant code path is intercepted. Every other call does not go into the default implementation of StdJDBCDelegate, but immediately delegates to the wrapped implementation. This provides more assurance that the wrapped class isn't being modified unnecessarily. From my review of the child classes, it does appear this isn't needed, but I wanted to be sure. It also future-proofs against overrides that I can't see yet.

As far as the hacker class, yes it is to provide a transparent, automatic solution. Without it, it would be pretty hard to dig up the needed configurations needed to solve this problem, exactly because this probably only affects a handful of users (if any). Google would probably never uncover this thread.

@li-ygerchikov - To be clear, this issue does not affect my organization, and I'm pretty much done trying to get this change in. I created a mess and I am simply trying to clean it up. I believe I have done my part. The committers of this project can either do their part to clean up this mess (which was included in 2 releases - not entirely my doing) or close this PR and accept this issue. Personally, I do find it theoretically possible that an organization might have grabbed one of the affected versions, locked into it, and might be deploying it to Prod as we speak. They might not discover the hole they are digging themselves into for another year. That possibility is why I put a significant amount of effort into fixing this issue.